In [1]:
import numpy as np
import pandas as pd 

In [2]:
# pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [3]:
df = pd.read_parquet("../../../Data/LLCP2022.parquet")

In [4]:
df

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE  \
0          1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
1          1.0     1.0  b'02042022'  b'02'  b'04'  b'2022'    1100.0   
2          1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   
3          1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
4          1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   
...        ...     ...          ...    ...    ...      ...       ...   
445127    78.0    11.0  b'12192022'  b'12'  b'19'  b'2022'    1100.0   
445128    78.0    11.0  b'12212022'  b'12'  b'21'  b'2022'    1100.0   
445129    78.0    11.0  b'11292022'  b'11'  b'29'  b'2022'    1100.0   
445130    78.0    11.0  b'12082022'  b'12'  b'08'  b'2022'    1100.0   
445131    78.0    11.0  b'12142022'  b'12'  b'14'  b'2022'    1100.0   

                SEQNO          _PSU  CTELENM1  PVTRESD1  COLGHOUS  STATERE1  \
0       b'2022000001'  2.022000e+09       1.0       1.0       NaN       1.0   
1       b'2022000002'  2.022000e+09       1.0       1.0       NaN       1.0   
2       b'2022000003'  2.022000e+09       1.0       1.0       NaN       1.0   
3       b'2022000004'  2.022000e+09       1.0       1.0       NaN       1.0   
4       b'2022000005'  2.022000e+09       1.0       1.0       NaN       1.0   
...               ...           ...       ...       ...       ...       ...   
445127  b'2022001527'  2.022002e+09       NaN       NaN       NaN       NaN   
445128  b'2022001528'  2.022002e+09       NaN       NaN       NaN       NaN   
445129  b'2022001529'  2.022002e+09       NaN       NaN       NaN       NaN   
445130  b'2022001530'  2.022002e+09       NaN       NaN       NaN       NaN   
445131  b'2022001531'  2.022002e+09       NaN       NaN       NaN       NaN   

        CELPHON1  LADULT1  COLGSEX1  NUMADULT  LANDSEX1  NUMMEN  NUMWOMEN  \
0            2.0      1.0       NaN       2.0       NaN     1.0       1.0   
1            2.0      1.0       NaN       2.0       NaN     1.0       1.0   
2            2.0      1.0       NaN       1.0       2.0     NaN       NaN   
3            2.0      1.0       NaN       3.0       NaN     2.0       1.0   
4            2.0      1.0       NaN       2.0       NaN     1.0       1.0   
...          ...      ...       ...       ...       ...     ...       ...   
445127       NaN      NaN       NaN       NaN       NaN     NaN       NaN   
445128       NaN      NaN       NaN       NaN       NaN     NaN       NaN   
445129       NaN      NaN       NaN       NaN       NaN     NaN       NaN   
445130       NaN      NaN       NaN       NaN       NaN     NaN       NaN   
445131       NaN      NaN       NaN       NaN       NaN     NaN       NaN   

        RESPSLCT  SAFETIME  CTELNUM1  CELLFON5  CADULT1  CELLSEX1  PVTRESD3  \
0            2.0       NaN       NaN       NaN      NaN       NaN       NaN   
1            2.0       NaN       NaN       NaN      NaN       NaN       NaN   
2            NaN       NaN       NaN       NaN      NaN       NaN       NaN   
3            2.0       NaN       NaN       NaN      NaN       NaN       NaN   
4            2.0       NaN       NaN       NaN      NaN       NaN       NaN   
...          ...       ...       ...       ...      ...       ...       ...   
445127       NaN       1.0       1.0       1.0      1.0       2.0       1.0   
445128       NaN       1.0       1.0       1.0      1.0       2.0       1.0   
445129       NaN       1.0       1.0       1.0      1.0       2.0       1.0   
445130       NaN       1.0       1.0       1.0      1.0       1.0       1.0   
445131       NaN       1.0       1.0       1.0      1.0       1.0       1.0   

        CCLGHOUS  CSTATE1  LANDLINE  HHADULT  SEXVAR  GENHLTH  PHYSHLTH  \
0            NaN      NaN       NaN      NaN     2.0      2.0      88.0   
1            NaN      NaN       NaN      NaN     2.0      1.0      88.0   
2            NaN      NaN       NaN      NaN     2.0      2.0       2.0   
3            NaN    

In [5]:
df.replace(to_replace='None', value=np.nan, inplace=True)
df = df.dropna(axis=1, how='all')

In [6]:
missing_percentage = (df.isnull().sum() / len(df)) * 100
columns_with_missing_values = missing_percentage[missing_percentage > 50]

result_df = pd.DataFrame({
    'Column': columns_with_missing_values.index,
    'Percentage_of_Missing_Values': columns_with_missing_values.values.round(2)
})
result_df.sort_values(by= 'Percentage_of_Missing_Values', ascending=False)

Column  Percentage_of_Missing_Values
2    COLGHOUS                        100.00
6    COLGSEX1                        100.00
12   CCLGHOUS                         99.70
85   CSRVCTL2                         99.65
60   HPVADSHT                         99.62
..        ...                           ...
18     HADMAM                         50.17
169  _CLNSCP1                         50.14
176  _CRCREC2                         50.11
170  _SGMSCP1                         50.08
171  _SGMS101                         50.08

[184 rows x 2 columns]

In [7]:
df = df.drop(columns = columns_with_missing_values.index, axis=1)
df

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE  \
0          1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
1          1.0     1.0  b'02042022'  b'02'  b'04'  b'2022'    1100.0   
2          1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   
3          1.0     1.0  b'02032022'  b'02'  b'03'  b'2022'    1100.0   
4          1.0     1.0  b'02022022'  b'02'  b'02'  b'2022'    1100.0   
...        ...     ...          ...    ...    ...      ...       ...   
445127    78.0    11.0  b'12192022'  b'12'  b'19'  b'2022'    1100.0   
445128    78.0    11.0  b'12212022'  b'12'  b'21'  b'2022'    1100.0   
445129    78.0    11.0  b'11292022'  b'11'  b'29'  b'2022'    1100.0   
445130    78.0    11.0  b'12082022'  b'12'  b'08'  b'2022'    1100.0   
445131    78.0    11.0  b'12142022'  b'12'  b'14'  b'2022'    1100.0   

                SEQNO          _PSU  SAFETIME  CTELNUM1  CELLFON5  CADULT1  \
0       b'2022000001'  2.022000e+09       NaN       NaN       NaN      NaN   
1       b'2022000002'  2.022000e+09       NaN       NaN       NaN      NaN   
2       b'2022000003'  2.022000e+09       NaN       NaN       NaN      NaN   
3       b'2022000004'  2.022000e+09       NaN       NaN       NaN      NaN   
4       b'2022000005'  2.022000e+09       NaN       NaN       NaN      NaN   
...               ...           ...       ...       ...       ...      ...   
445127  b'2022001527'  2.022002e+09       1.0       1.0       1.0      1.0   
445128  b'2022001528'  2.022002e+09       1.0       1.0       1.0      1.0   
445129  b'2022001529'  2.022002e+09       1.0       1.0       1.0      1.0   
445130  b'2022001530'  2.022002e+09       1.0       1.0       1.0      1.0   
445131  b'2022001531'  2.022002e+09       1.0       1.0       1.0      1.0   

        CELLSEX1  PVTRESD3  CSTATE1  LANDLINE  HHADULT  SEXVAR  GENHLTH  \
0            NaN       NaN      NaN       NaN      NaN     2.0      2.0   
1            NaN       NaN      NaN       NaN      NaN     2.0      1.0   
2            NaN       NaN      NaN       NaN      NaN     2.0      2.0   
3            NaN       NaN      NaN       NaN      NaN     2.0      1.0   
4            NaN       NaN      NaN       NaN      NaN     2.0      4.0   
...          ...       ...      ...       ...      ...     ...      ...   
445127       2.0       1.0      1.0       2.0      5.0     2.0      3.0   
445128       2.0       1.0      1.0       1.0      2.0     2.0      1.0   
445129       2.0       1.0      1.0       2.0      1.0     2.0      5.0   
445130       1.0       1.0      1.0       2.0      2.0     1.0      2.0   
445131       1.0       1.0      1.0       2.0      1.0     1.0      2.0   

        PHYSHLTH  MENTHLTH  POORHLTH  PRIMINSR  PERSDOC3  MEDCOST1  CHECKUP1  \
0           88.0      88.0       NaN      99.0       1.0       2.0       1.0   
1           88.0      88.0       NaN       3.0       2.0       2.0       8.0   
2            2.0       3.0       2.0       1.0       1.0       2.0       1.0   
3           88.0      88.0       NaN      99.0       1.0       2.0       1.0   
4            2.0      88.0      88.0       7.0       2.0       2.0       1.0   
...          ...       ...       ...       ...       ...       ...       ...   
445127      88.0       3.0      30.0       3.0       3.0       2.0       2.0   
445128       2.0       2.0      88.0       1.0       2.0       2.0       1.0   
445129      30.0      30.0      30.0      88.0       3.0       1.0       4.0   
445130      88.0      88.0       NaN       3.0       2.0       2.0       1.0   
445131      88.0       1.0      88.0      88.0       3.0       1.0       8.0   

        EXERANY2  SLEPTIM1  LASTDEN4  RMVTETH4  CVDINFR4  CVDCRHD4  CVDSTRK3  \
0            2.0       8.0       NaN       NaN       2.0       2.0       2.0   
1            2.0       6.0       NaN       NaN       2.0       2.0       2.0   
2            1.0       5.0       NaN       NaN       2.0       2.0       2.0   
3            1.0       7

In [8]:
missing_percentage = (df.isnull().sum() / len(df)) * 100
columns_with_missing_values = missing_percentage[missing_percentage >= 0]

result_df = pd.DataFrame({
    'Column': columns_with_missing_values.index,
    'Percentage_of_Missing_Values': columns_with_missing_values.values.round(2)
})
# result_df.sort_values(by= 'Percentage_of_Missing_Values', ascending=False)

In [9]:
df = df[result_df[result_df['Percentage_of_Missing_Values'] < 10.0].Column.value_counts().index].dropna()
invalid_columns = ['CVDCRHD4', 'CVDINFR4', 'LCSCTSC1']
# These columns are being removed because they are asking the same information as the target and should not skew the weights
df = df.drop(invalid_columns ,axis =1)
df

_STATE  _MRACE2  _DRDXAR2  _ASTHMS1  _CASTHM1  _LTASTH1  _MICHD  \
342        1.0      1.0       1.0       3.0       1.0       1.0     2.0   
343        1.0      1.0       1.0       3.0       1.0       1.0     2.0   
344        1.0      1.0       2.0       3.0       1.0       1.0     2.0   
345        1.0      1.0       1.0       3.0       1.0       1.0     2.0   
346        1.0      1.0       1.0       3.0       1.0       1.0     2.0   
...        ...      ...       ...       ...       ...       ...     ...   
435819    56.0      1.0       2.0       1.0       2.0       2.0     2.0   
435820    56.0      1.0       2.0       3.0       1.0       1.0     1.0   
435822    56.0      1.0       2.0       3.0       1.0       1.0     2.0   
435823    56.0      1.0       2.0       3.0       1.0       1.0     1.0   
435824    56.0      1.0       2.0       3.0       1.0       1.0     2.0   

        _DENVST3  _EXTETH3  _TOTINDA  _HCVU652  _HLTHPLN  _MENT14D  _PHYS14D  \
342          2.0       1.0       1.0       9.0       1.0       1.0       2.0   
343          2.0       1.0       1.0       9.0       1.0       1.0       1.0   
344          2.0       2.0       2.0       9.0       1.0       1.0       3.0   
345          2.0       2.0       2.0       9.0       1.0       1.0       1.0   
346          2.0       1.0       1.0       9.0       1.0       1.0       2.0   
...          ...       ...       ...       ...       ...       ...       ...   
435819       1.0       2.0       1.0       1.0       1.0       1.0       1.0   
435820       1.0       2.0       2.0       9.0       1.0       2.0       2.0   
435822       1.0       1.0       1.0       9.0       1.0       1.0       9.0   
435823       9.0       1.0       1.0       1.0       1.0       1.0       2.0   
435824       1.0       1.0       1.0       1.0       1.0       2.0       1.0   

        _RFHLTH     _LLCPWT    _LLCPWT2  _DUALUSE  _IMPRACE   _WT2RAKE  \
342         1.0  957.778505  813.918517       1.0       1.0  74.836832   
343         1.0  772.369813  813.918517       1.0       1.0  74.836832   
344         2.0  477.761050  813.918517       1.0       1.0  74.836832   
345         1.0  463.222081  813.918517       1.0       1.0  74.836832   
346         2.0  394.148445  406.959258       1.0       1.0  37.418416   
...         ...         ...         ...       ...       ...        ...   
435819      1.0  224.792895  144.397206       9.0       1.0   5.992375   
435820      2.0   70.898673  144.397206       9.0       1.0   5.992375   
435822      1.0   54.265519   99.449356       2.0       1.0   5.992375   
435823      1.0  130.285109  144.397206       9.0       1.0   5.992375   
435824      1.0  105.276935   99.449356       2.0       1.0   5.992375   

        _RAWRAKE     _STRWT    _STSTR  _URBSTAT  _METSTAT  _PRACE2  _HISPANC  \
342          2.0  37.418416   11011.0       1.0       2.0      1.0       2.0   
343          2.0  37.418416   11011.0       1.0       1.0      1.0       2.0   
344          2.0  37.418416   11011.0       1.0       1.0      1.0       2.0   
345          2.0  37.418416   11011.0       1.0       1.0      1.0       2.0   
346          1.0  37.418416   11011.0       1.0       1.0      1.0       2.0   
...          ...        ...       ...       ...       ...      ...       ...   
435819       1.0   5.992375  562011.0       2.0       2.0      1.0       2.0   
435820       1.0   5.992375  562012.0       1.0       1.0      1.0       2.0   
435822       1.0   5.992375  562011.0       1.0       2.0      1.0       2.0   
435823       1.0   5.992375  562011.0       2.0       2.0      1.0       2.0   
435824       1.0   5.992375  562012.0       1.0       2.0      1.0       2.0   

        QSTVER  _RACE1      _DRNKWK2  _RFBING6      DROCDY4_  DRNKANY6  \
342       10.0     1.0  5.397605e-79       1.0  5.397605e-79       2.0   
343       10.0     1.0  5.397605e-79       1.0  5.397605e-79       2.0   
344       10.0     1.0  5.397605e-79       1.0  5.397605e-79       2.0   


In [10]:
target = df._MICHD
df = df.drop('_MICHD', axis=1)
target

342       2.0
343       2.0
344       2.0
345       2.0
346       2.0
         ... 
435819    2.0
435820    1.0
435822    2.0
435823    1.0
435824    2.0
Name: _MICHD, Length: 355111, dtype: float64

In [11]:
# value_counts_per_column = df.apply(lambda x: x.value_counts())

# # Step 2: Check for columns where the unique values are 1.0 and 2.0 only
# columns_with_1_and_2 = value_counts_per_column.columns[(value_counts_per_column.loc[1.0] + value_counts_per_column.loc[2.0]).eq(value_counts_per_column.sum(axis=0))]

# # Step 3: Display the columns
# print("Columns with value counts of 1.0 and 2.0 only:")
# print(columns_with_1_and_2)

In [12]:
# value_counts_per_column = df.apply(lambda x: x.value_counts())
# columns_with_4_or_less_unique = value_counts_per_column.columns[value_counts_per_column.nunique() <= 4]

# print(columns_with_4_or_less_unique)

In [13]:
# for col in columns_with_4_or_less_unique:
#     print(df[col].value_counts())

In [11]:
unique_counts_df = pd.DataFrame({
    'Column Name': df.columns,
    'Number of Unique Values': df.nunique()
})
unique_counts_df.sort_values(by='Number of Unique Values', ascending=False)

Column Name  Number of Unique Values
_LLCPWT      _LLCPWT                   211924
SEQNO          SEQNO                    24601
_PSU            _PSU                    24601
_LLCPWT2    _LLCPWT2                    11893
_WT2RAKE    _WT2RAKE                     5020
...              ...                      ...
_RFSMOK3    _RFSMOK3                        2
SMOKE100    SMOKE100                        2
_SEX            _SEX                        2
SEXVAR        SEXVAR                        2
QSTLANG      QSTLANG                        2

[103 rows x 2 columns]

In [15]:
# colsW10OrLessGroups = unique_counts_df[unique_counts_df['Number of Unique Values'] <10]['Column Name']
# for col in colsW10OrLessGroups:
#     print(df[col].value_counts())

# print("Number of Columns with Less than 10 unique values:  ", len(colsW10OrLessGroups))

In [16]:
# colsW10To20Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 10) & (unique_counts_df['Number of Unique Values'] < 20)]['Column Name']

# for col in colsW10To20Groups:
#     print(df[col].value_counts())

# print("Number of columns with 10 - 20 unique values:   ", len(colsW10To20Groups))

So CHILDREN and _STATE are the two categorical values and everything over 20 unique values is a numerical value. 

In [17]:
# colsW20To30Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 20) & (unique_counts_df['Number of Unique Values'] < 30)]['Column Name']

# for col in colsW20To30Groups:
#     print(df[col].value_counts())
# print("Number of columns with 20 - 30 unique values:   ", len(colsW20To30Groups))

In [18]:
# colsW30To40Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 30) & (unique_counts_df['Number of Unique Values'] < 40)]['Column Name']

# for col in colsW30To40Groups:
#     print(df[col].value_counts())
# print("Number of columns with 30 - 40 unique values:   ", len(colsW30To40Groups))

In [19]:
# colsW40To50Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 40) & (unique_counts_df['Number of Unique Values'] < 50)]['Column Name']

# for col in colsW40To50Groups:
#     print(df[col].value_counts())
# print("Number of columns with 40 - 50 unique values:   ", len(colsW40To50Groups))

In [20]:
# colsW50To60Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 50) & (unique_counts_df['Number of Unique Values'] < 60)]['Column Name']

# for col in colsW50To60Groups:
#     print(df[col].value_counts())
# print("Number of columns with 50 - 60 unique values:   ", len(colsW50To60Groups))

In [21]:
# colsW60To70Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 60) & (unique_counts_df['Number of Unique Values'] < 70)]['Column Name']

# for col in colsW60To70Groups:
#     print(df[col].value_counts())
# print("Number of columns with 60 - 70 unique values:   ", len(colsW60To70Groups))

In [22]:
# colsW70To1000Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 70) & (unique_counts_df['Number of Unique Values'] < 1000)]['Column Name']

# for col in colsW70To1000Groups:
#     print(df[col].value_counts())
# print("Number of columns with 70 - 1000 unique values:   ", len(colsW70To1000Groups))

In [23]:
# colsW1000To80Groups = unique_counts_df[(unique_counts_df['Number of Unique Values'] >= 1000) & (unique_counts_df['Number of Unique Values'] < 60000)]['Column Name']

# for col in colsW1000To80Groups:
#     print(df[col].value_counts())
# print("Number of columns with 1000 - 80 unique values:   ", len(colsW1000To80Groups))

In [12]:
unique_counts_df[(unique_counts_df['Number of Unique Values'] <=20)]['Column Name']

_MRACE2      _MRACE2
_DRDXAR2    _DRDXAR2
_ASTHMS1    _ASTHMS1
_CASTHM1    _CASTHM1
_LTASTH1    _LTASTH1
              ...   
MARITAL      MARITAL
DIABETE4    DIABETE4
HAVARTH4    HAVARTH4
CHCKDNY2    CHCKDNY2
_RFDRHV8    _RFDRHV8
Name: Column Name, Length: 80, dtype: object

In [13]:
categorical_columns = ["CHILDREN", "_STATE"]
filtered_columns = list(unique_counts_df[(unique_counts_df['Number of Unique Values'] <= 20)]['Column Name'])

# # Ensure that filtered_columns are present in df
# filtered_columns = [col for col in filtered_columns if col in df.columns]

df_categorical = df[categorical_columns + filtered_columns]
# df_categorical = df_categorical.drop('_MICHD', axis=1)


In [14]:
df_categorical

CHILDREN  _STATE  _MRACE2  _DRDXAR2  _ASTHMS1  _CASTHM1  _LTASTH1  \
342         88.0     1.0      1.0       1.0       3.0       1.0       1.0   
343         88.0     1.0      1.0       1.0       3.0       1.0       1.0   
344          1.0     1.0      1.0       2.0       3.0       1.0       1.0   
345         88.0     1.0      1.0       1.0       3.0       1.0       1.0   
346         88.0     1.0      1.0       1.0       3.0       1.0       1.0   
...          ...     ...      ...       ...       ...       ...       ...   
435819      88.0    56.0      1.0       2.0       1.0       2.0       2.0   
435820      88.0    56.0      1.0       2.0       3.0       1.0       1.0   
435822      88.0    56.0      1.0       2.0       3.0       1.0       1.0   
435823      88.0    56.0      1.0       2.0       3.0       1.0       1.0   
435824      88.0    56.0      1.0       2.0       3.0       1.0       1.0   

        _DENVST3  _EXTETH3  _TOTINDA  _HCVU652  _HLTHPLN  _MENT14D  _PHYS14D  \
342          2.0       1.0       1.0       9.0       1.0       1.0       2.0   
343          2.0       1.0       1.0       9.0       1.0       1.0       1.0   
344          2.0       2.0       2.0       9.0       1.0       1.0       3.0   
345          2.0       2.0       2.0       9.0       1.0       1.0       1.0   
346          2.0       1.0       1.0       9.0       1.0       1.0       2.0   
...          ...       ...       ...       ...       ...       ...       ...   
435819       1.0       2.0       1.0       1.0       1.0       1.0       1.0   
435820       1.0       2.0       2.0       9.0       1.0       2.0       2.0   
435822       1.0       1.0       1.0       9.0       1.0       1.0       9.0   
435823       9.0       1.0       1.0       1.0       1.0       1.0       2.0   
435824       1.0       1.0       1.0       1.0       1.0       2.0       1.0   

        _RFHLTH  _DUALUSE  _IMPRACE  _RAWRAKE  _URBSTAT  _METSTAT  _PRACE2  \
342         1.0       1.0       1.0       2.0       1.0       2.0      1.0   
343         1.0       1.0       1.0       2.0       1.0       1.0      1.0   
344         2.0       1.0       1.0       2.0       1.0       1.0      1.0   
345         1.0       1.0       1.0       2.0       1.0       1.0      1.0   
346         2.0       1.0       1.0       1.0       1.0       1.0      1.0   
...         ...       ...       ...       ...       ...       ...      ...   
435819      1.0       9.0       1.0       1.0       2.0       2.0      1.0   
435820      2.0       9.0       1.0       1.0       1.0       1.0      1.0   
435822      1.0       2.0       1.0       1.0       1.0       2.0      1.0   
435823      1.0       9.0       1.0       1.0       2.0       2.0      1.0   
435824      1.0       2.0       1.0       1.0       1.0       2.0      1.0   

        _HISPANC  QSTVER  _RACE1  _RFBING6  DRNKANY6  _SMOKGRP  _CURECI2  \
342          2.0    10.0     1.0       1.0       2.0       3.0       1.0   
343          2.0    10.0     1.0       1.0       2.0       3.0       1.0   
344          2.0    10.0     1.0       1.0       2.0       3.0       1.0   
345          2.0    10.0     1.0       2.0       1.0       3.0       1.0   
346          2.0    10.0     1.0       1.0       2.0       4.0       1.0   
...          ...     ...     ...       ...       ...       ...       ...   
435819       2.0    20.0     1.0       1.0       1.0       3.0       1.0   
435820       2.0    20.0     1.0       1.0       2.0       1.0       1.0   
435822       2.0    20.0     1.0       1.0       1.0       4.0       1.0   
435823       2.0    20.0     1.0       1.0       1.0       3.0       1.0   
435824       2.0    20.0     1.0       1.0       1.0       4.0       1.0   

        _RFSMOK3  _SMOKER3  _INCOMG1  _EDUCAG  _CHLDCNT  _RFBMI5  _AGE_G  \
342          1.0       3.0       6.0      3.0       1.0      2.0     6.0   
343          1.0       3.0       5.0      3.0       1.0      2.0     6.0   
344          1.0       3.0       5.0      4.

In [15]:
cols = list(df_categorical.columns)
mask = unique_counts_df['Column Name'].isin(cols)
filtered_rows = unique_counts_df[mask]
filtered_rows.sort_values(by = ['Number of Unique Values'], ascending = False)

Column Name  Number of Unique Values
_STATE        _STATE                       51
CHILDREN    CHILDREN                       29
_AGEG5YR    _AGEG5YR                       14
PRIMINSR    PRIMINSR                       13
INCOME3      INCOME3                       13
...              ...                      ...
DISPCODE    DISPCODE                        2
IYEAR          IYEAR                        2
_RFSMOK3    _RFSMOK3                        2
SMOKE100    SMOKE100                        2
SEXVAR        SEXVAR                        2

[82 rows x 2 columns]

In [16]:
df_categorical = df_categorical.astype('category')
df_categorical.dtypes

CHILDREN    category
_STATE      category
_MRACE2     category
_DRDXAR2    category
_ASTHMS1    category
              ...   
MARITAL     category
DIABETE4    category
HAVARTH4    category
CHCKDNY2    category
_RFDRHV8    category
Length: 82, dtype: object

In [17]:
numerical_columns = list(set(df.columns) - set(df_categorical.columns))
df_numerical = df[numerical_columns]
df_numerical = df_numerical.drop(['IDAY', 'IDATE', 'SEQNO'], axis=1)
df_numerical

_PSU  HTIN4     _STRWT  ALCDAY4  MENTHLTH    _LLCPWT2  \
342     2.022000e+09   63.0  37.418416    888.0      88.0  813.918517   
343     2.022000e+09   70.0  37.418416    888.0      88.0  813.918517   
344     2.022000e+09   72.0  37.418416    888.0      88.0  813.918517   
345     2.022000e+09   73.0  37.418416    104.0      88.0  813.918517   
346     2.022000e+09   67.0  37.418416    888.0      88.0  406.959258   
...              ...    ...        ...      ...       ...         ...   
435819  2.022004e+09   62.0   5.992375    203.0      88.0  144.397206   
435820  2.022004e+09   70.0   5.992375    888.0       2.0  144.397206   
435822  2.022004e+09   61.0   5.992375    210.0      88.0   99.449356   
435823  2.022004e+09   70.0   5.992375    202.0      88.0  144.397206   
435824  2.022004e+09   64.0   5.992375    205.0       5.0   99.449356   

          WTKG3      DROCDY4_  HEIGHT3  _AGE80    _STSTR  WEIGHT2     _LLCPWT  \
342      7167.0  5.397605e-79    503.0    68.0   11011.0    158.0  957.778505   
343      9525.0  5.397605e-79    510.0    71.0   11011.0    210.0  772.369813   
344      8845.0  5.397605e-79    600.0    72.0   11011.0    195.0  477.761050   
345     10886.0  5.700000e+01    601.0    78.0   11011.0    240.0  463.222081   
346      9072.0  5.397605e-79    507.0    80.0   11011.0    200.0  394.148445   
...         ...           ...      ...     ...       ...      ...         ...   
435819   5670.0  1.000000e+01    502.0    42.0  562011.0    125.0  224.792895   
435820   8482.0  5.397605e-79    510.0    68.0  562012.0    187.0   70.898673   
435822   6804.0  3.300000e+01    501.0    80.0  562011.0    150.0   54.265519   
435823   9344.0  7.000000e+00    510.0    56.0  562011.0    206.0  130.285109   
435824  10206.0  1.700000e+01    504.0    59.0  562012.0    225.0  105.276935   

        SLEPTIM1  PHYSHLTH      _DRNKWK2   HTM4   _WT2RAKE  
342          9.0       4.0  5.397605e-79  160.0  74.836832  
343          6.0      88.0  5.397605e-79  178.0  74.836832  
344          8.0      30.0  5.397605e-79  183.0  74.836832  
345          8.0      88.0  8.000000e+02  185.0  74.836832  
346          9.0       5.0  5.397605e-79  170.0  37.418416  
...          ...       ...           ...    ...        ...  
435819       8.0      88.0  2.100000e+02  157.0   5.992375  
435820       6.0       2.0  5.397605e-79  178.0   5.992375  
435822       7.0      77.0  2.330000e+02  155.0   5.992375  
435823       8.0       7.0  1.400000e+02  178.0   5.992375  
435824       6.0      88.0  2.330000e+02  163.0   5.992375  

[355111 rows x 18 columns]

In [18]:
df_numerical.dtypes.sort_values()

_PSU        float64
_DRNKWK2    float64
PHYSHLTH    float64
SLEPTIM1    float64
_LLCPWT     float64
WEIGHT2     float64
_STSTR      float64
_AGE80      float64
HEIGHT3     float64
DROCDY4_    float64
WTKG3       float64
_LLCPWT2    float64
MENTHLTH    float64
ALCDAY4     float64
_STRWT      float64
HTIN4       float64
HTM4        float64
_WT2RAKE    float64
dtype: object

In [19]:
# Get the total number of columns
total_columns = df.shape[1]

# Calculate the indices to split the columns into eight equal parts
eighth1 = total_columns // 8
eighth2 = 2 * total_columns // 8
eighth3 = 3 * total_columns // 8
eighth4 = 4 * total_columns // 8
eighth5 = 5 * total_columns // 8
eighth6 = 6 * total_columns // 8
eighth7 = 7 * total_columns // 8

# Slice the DataFrame to get the eight equal parts
part1 = df_categorical.iloc[:, :eighth1]
part2 = df_categorical.iloc[:, eighth1:eighth2]
part3 = df_categorical.iloc[:, eighth2:eighth3]
part4 = df_categorical.iloc[:, eighth3:eighth4]
part5 = df_categorical.iloc[:, eighth4:eighth5]
part6 = df_categorical.iloc[:, eighth5:eighth6]
part7 = df_categorical.iloc[:, eighth6:eighth7]
parts = [part1, part2, part3, part4, part5, part6, part7]

print(part1.shape, part2.shape, part3.shape, part4.shape, part5.shape, part6.shape, part7.shape)
total_col = 0
for part in parts:
    total_col += part.shape[1]
print(total_col, df_categorical.shape)

(355111, 12) (355111, 13) (355111, 13) (355111, 13) (355111, 13) (355111, 13) (355111, 5)
82 (355111, 82)


In [20]:
index = 0
for part in parts:
    part = pd.get_dummies(part)
    print(part.shape)
    parts[index] = part
    index+=1



(355111, 116)
(355111, 67)
(355111, 65)
(355111, 70)
(355111, 70)
(355111, 73)
(355111, 22)


In [21]:
df_categorical = pd.concat(parts, axis=1)
df_categorical.shape

(355111, 483)

In [22]:
df_categorical.to_parquet("categoricalWithDummies.parquet")

The shape is (355111, 495) of the categorical with dummies.

In [23]:
df_numerical = (df_numerical - df_numerical.mean()) / df_numerical.std()

In [24]:
df_numerical.head()

_PSU     HTIN4    _STRWT   ALCDAY4  MENTHLTH  _LLCPWT2     WTKG3  \
342 -1.135672 -0.997923  0.036216  1.072327  0.800439  0.267583 -0.547620   
343 -1.135447  0.672952  0.036216  1.072327  0.800439  0.267583  0.550557   
344 -1.135223  1.150345  0.036216  1.072327  0.800439  0.267583  0.233864   
345 -1.134998  1.389041  0.036216 -1.123638  0.800439  0.267583  1.184407   
346 -1.134774 -0.043137  0.036216  1.072327  0.800439 -0.213051  0.339584   

     DROCDY4_   HEIGHT3    _AGE80    _STSTR   WEIGHT2   _LLCPWT  SLEPTIM1  \
342 -0.286710 -0.458355  0.725882 -1.881422 -0.109000  0.350352  0.185031   
343 -0.286710 -0.278831  0.894640 -1.881422  0.018999  0.179652 -0.239291   
344 -0.286710  2.029329  0.950892 -1.881422 -0.017924 -0.091585  0.043590   
345  0.313408  2.054975  1.288407 -1.881422  0.092845 -0.104970  0.043590   
346 -0.286710 -0.355770  1.400912 -1.881422 -0.005616 -0.168564  0.185031   

     PHYSHLTH  _DRNKWK2      HTM4  _WT2RAKE  
342 -1.447494 -0.152674 -0.998139  0.633109  
343  0.774994 -0.152674  0.689488  0.633109  
344 -0.759581 -0.152674  1.158273  0.633109  
345  0.774994 -0.089098  1.345787  0.633109  
346 -1.421035 -0.152674 -0.060568 -0.008318

In [25]:
df_numerical.to_parquet("numericalStandardized2022.parquet")

In [26]:
test_num = pd.read_parquet("numericalStandardized2022.parquet")
test_num

_PSU     HTIN4    _STRWT   ALCDAY4  MENTHLTH  _LLCPWT2     WTKG3  \
342    -1.135672 -0.997923  0.036216  1.072327  0.800439  0.267583 -0.547620   
343    -1.135447  0.672952  0.036216  1.072327  0.800439  0.267583  0.550557   
344    -1.135223  1.150345  0.036216  1.072327  0.800439  0.267583  0.233864   
345    -1.134998  1.389041  0.036216 -1.123638  0.800439  0.267583  1.184407   
346    -1.134774 -0.043137  0.036216  1.072327  0.800439 -0.213051  0.339584   
...          ...       ...       ...       ...       ...       ...       ...   
435819 -0.284347 -1.236620 -0.563271 -0.846341  0.800439 -0.523146 -1.244809   
435820 -0.284122  0.672952 -0.563271  1.072327 -1.456003 -0.523146  0.064807   
435822 -0.283673 -1.475316 -0.563271 -0.826734  0.800439 -0.576231 -0.716678   
435823 -0.283449  0.672952 -0.563271 -0.849142  0.800439 -0.523146  0.466261   
435824 -0.283224 -0.759227 -0.563271 -0.840739 -1.377290 -0.576231  0.867715   

        DROCDY4_   HEIGHT3    _AGE80    _STSTR   WEIGHT2   _LLCPWT  SLEPTIM1  \
342    -0.286710 -0.458355  0.725882 -1.881422 -0.109000  0.350352  0.185031   
343    -0.286710 -0.278831  0.894640 -1.881422  0.018999  0.179652 -0.239291   
344    -0.286710  2.029329  0.950892 -1.881422 -0.017924 -0.091585  0.043590   
345     0.313408  2.054975  1.288407 -1.881422  0.092845 -0.104970  0.043590   
346    -0.286710 -0.355770  1.400912 -1.881422 -0.005616 -0.168564  0.185031   
...          ...       ...       ...       ...       ...       ...       ...   
435819 -0.181426 -0.484001 -0.736683  1.645614 -0.190231 -0.324484  0.043590   
435820 -0.286710 -0.278831  0.725882  1.645620 -0.037616 -0.466170 -0.239291   
435822  0.060727 -0.509647  1.400912  1.645614 -0.128692 -0.481484 -0.097850   
435823 -0.213011 -0.278831  0.050852  1.645614  0.009153 -0.411495  0.043590   
435824 -0.107727 -0.432709  0.219610  1.645620  0.055922 -0.434519 -0.239291   

        PHYSHLTH  _DRNKWK2      HTM4  _WT2RAKE  
342    -1.447494 -0.152674 -0.998139  0.633109  
343     0.774994 -0.152674  0.689488  0.633109  
344    -0.759581 -0.152674  1.158273  0.633109  
345     0.774994 -0.089098  1.345787  0.633109  
346    -1.421035 -0.152674 -0.060568 -0.008318  
...          ...       ...       ...       ...  
435819  0.774994 -0.135985 -1.279410 -0.547023  
435820 -1.500410 -0.152674  0.689488 -0.547023  
435822  0.483954 -0.134157 -1.466924 -0.547023  
435823 -1.368119 -0.141548  0.689488 -0.547023  
435824  0.774994 -0.134157 -0.716868 -0.547023  

[355111 rows x 18 columns]

In [27]:
test_cat = pd.read_parquet("categoricalWithDummies.parquet")
test_cat

CHILDREN_1.0  CHILDREN_2.0  CHILDREN_3.0  CHILDREN_4.0  CHILDREN_5.0  \
342                0             0             0             0             0   
343                0             0             0             0             0   
344                1             0             0             0             0   
345                0             0             0             0             0   
346                0             0             0             0             0   
...              ...           ...           ...           ...           ...   
435819             0             0             0             0             0   
435820             0             0             0             0             0   
435822             0             0             0             0             0   
435823             0             0             0             0             0   
435824             0             0             0             0             0   

        CHILDREN_6.0  CHILDREN_7.0  CHILDREN_8.0  CHILDREN_9.0  CHILDREN_10.0  \
342                0             0             0             0              0   
343                0             0             0             0              0   
344                0             0             0             0              0   
345                0             0             0             0              0   
346                0             0             0             0              0   
...              ...           ...           ...           ...            ...   
435819             0             0             0             0              0   
435820             0             0             0             0              0   
435822             0             0             0             0              0   
435823             0             0             0             0              0   
435824             0             0             0             0              0   

        CHILDREN_11.0  CHILDREN_12.0  CHILDREN_13.0  CHILDREN_14.0  \
342                 0              0              0              0   
343                 0              0              0              0   
344                 0              0              0              0   
345                 0              0              0              0   
346                 0              0              0              0   
...               ...            ...            ...            ...   
435819              0              0              0              0   
435820              0              0              0              0   
435822              0              0              0              0   
435823              0              0              0              0   
435824              0              0              0              0   

        CHILDREN_15.0  CHILDREN_17.0  CHILDREN_20.0  CHILDREN_21.0  \
342                 0              0              0              0   
343                 0              0              0              0   
344                 0              0              0              0   
345                 0              0              0              0   
346                 0              0              0              0   
...               ...            ...            ...            ...   
435819              0              0              0              0   
435820              0              0              0              0   
435822              0              0              0              0   
435823              0              0              0              0   
435824              0              0              0              0   

        CHILDREN_22.0  CHILDREN_23.0  CHILDREN_25.0  CHILDREN_30.0  \
342                 0              0              0              0   
343                 0              0              0              0   
344                 0              0              0              0   
345                 0              0              0              0   
346                 0              0      

In [28]:
df_proc = pd.concat([df_categorical, df_numerical, target], axis=1)
df_proc

CHILDREN_1.0  CHILDREN_2.0  CHILDREN_3.0  CHILDREN_4.0  CHILDREN_5.0  \
342                0             0             0             0             0   
343                0             0             0             0             0   
344                1             0             0             0             0   
345                0             0             0             0             0   
346                0             0             0             0             0   
...              ...           ...           ...           ...           ...   
435819             0             0             0             0             0   
435820             0             0             0             0             0   
435822             0             0             0             0             0   
435823             0             0             0             0             0   
435824             0             0             0             0             0   

        CHILDREN_6.0  CHILDREN_7.0  CHILDREN_8.0  CHILDREN_9.0  CHILDREN_10.0  \
342                0             0             0             0              0   
343                0             0             0             0              0   
344                0             0             0             0              0   
345                0             0             0             0              0   
346                0             0             0             0              0   
...              ...           ...           ...           ...            ...   
435819             0             0             0             0              0   
435820             0             0             0             0              0   
435822             0             0             0             0              0   
435823             0             0             0             0              0   
435824             0             0             0             0              0   

        CHILDREN_11.0  CHILDREN_12.0  CHILDREN_13.0  CHILDREN_14.0  \
342                 0              0              0              0   
343                 0              0              0              0   
344                 0              0              0              0   
345                 0              0              0              0   
346                 0              0              0              0   
...               ...            ...            ...            ...   
435819              0              0              0              0   
435820              0              0              0              0   
435822              0              0              0              0   
435823              0              0              0              0   
435824              0              0              0              0   

        CHILDREN_15.0  CHILDREN_17.0  CHILDREN_20.0  CHILDREN_21.0  \
342                 0              0              0              0   
343                 0              0              0              0   
344                 0              0              0              0   
345                 0              0              0              0   
346                 0              0              0              0   
...               ...            ...            ...            ...   
435819              0              0              0              0   
435820              0              0              0              0   
435822              0              0              0              0   
435823              0              0              0              0   
435824              0              0              0              0   

        CHILDREN_22.0  CHILDREN_23.0  CHILDREN_25.0  CHILDREN_30.0  \
342                 0              0              0              0   
343                 0              0              0              0   
344                 0              0              0              0   
345                 0              0              0              0   
346                 0              0      

In [29]:
duplicate_columns = df_proc.columns[df_proc.columns.duplicated()]
duplicate_columns

Index([], dtype='object')

In [30]:
df_proc.to_parquet("processed_Dummies_Stand_2022.parquet")